### PR TITLE
[Core] Factor out input preprocessing to a separate class

### PR DESCRIPTION
**Pull request by DarkLight1337**
Following up to #7258, this PR implements the suggestion by @robertgshaw2-neuralmagic to create a new class for prompt parsing. This helps keep the sync/async versions of the preprocessing code together so they can be updated together easily, while also slimming down the existing engine classes to focus on request processing.

### CODE REVIEW

File: vllm/inputs/preprocess.py [hunk 1]
```diff
@@ -0,0 +1,536 @@
...
+        Returns:
+
+        * prompt
+        * prompt_token_ids
+        * multi_modal_data
+        '''
+
+        parsed = parse_singleton_prompt(inputs)
+
+        if parsed["type"] == "str":
```
Comment (njhill) [inline vllm/inputs/preprocess.py:231]: have a `prompt_type = parsed["type"]` and reuse instead of repeated lookups? (same for `parsed["content"]`)

**Review by njhill — COMMENTED** [review]
Thanks @DarkLight1337, just some minor comments.

We may need to re-think some of this soon since we're hoping to get rid of AsyncLLMEngine altogether. I'm thinking we'll hopefully no longer need both sync and async versions of these things.

**Comment by DarkLight1337** [comment]
Thanks for the review! I'll address the detailed comments shortly.

> We may need to re-think some of this soon since we're hoping to get rid of AsyncLLMEngine altogether. I'm thinking we'll hopefully no longer need both sync and async versions of these things.

Even without `AsyncLLMEngine`, the preprocessing logic better fits inside `vllm.inputs` module, and it would reduce the bloat inside `LLMEngine` itself.

File: vllm/inputs/preprocess.py [hunk 2]
```diff
@@ -0,0 +1,536 @@
...
+    async def _extract_prompt_components_async(
+        self,
+        inputs: SingletonPromptInputs,
+        request_id: str,
+        lora_request: Optional[LoRARequest] = None,
+    ) -> PromptComponents:
+        """Async version of :meth:`_extract_prompt_components`."""
+        parsed = parse_singleton_prompt(inputs)
+
+        if parsed["type"] == "str":
```
[the remark this answers is not in the document]
Comment (DarkLight1337) [reply vllm/inputs/preprocess.py:265]: I tried this, and unfortunately this breaks the type checker's capability to narrow down the type. So I'm keeping this as is.

File: vllm/inputs/preprocess.py (hunk 1 above)
Comment (DarkLight1337) [reply vllm/inputs/preprocess.py:231]: As above.

**Comment by DarkLight1337** [comment]
I have finished addressing your other comments.

**Review by njhill — APPROVED** [review]

**Pull request merged**